apps/emqx_durable_storage/src/emqx_ds.erl

```diff
@@ -94,6 +97,32 @@
 %% Parsed topic filter.
 -type topic_filter() :: list(binary() | '+' | '#' | '').
 
+-type message() :: emqx_types:message().
+
+%% Message pattern.
+-type message(Pattern) :: emqx_types:message(Pattern).
+
+%% A batch of storage operations.
+-type batch() :: [operation()] | #dsbatch{}.
+
+-type operation() ::
+    %% Store a message.
+    message()
+    %% Delete a message.
+    %% Does nothing if the message does not exist.
+    | {delete, message(_)}.
```

Comment:
Since we decided to split out preconditions, why not do the same to deletes?

Reply:
...And supposedly should be part of storage-layer-level batches.

Reply:
Ok.

Reply:
Another question: why is it `{delete, #message{}}` and not `{delete, Key}`? Deletion by key is trivial for DS, deletion by message requires non-trivial matching.

Reply:
Nvm, answered my own question.